File: app/src/main/kotlin/org/wordpress/aztec/demo/PicassoImageLoader.java
package org.wordpress.aztec.demo;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.support.v4.util.ArrayMap;

import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import org.wordpress.aztec.AztecText;
import org.wordpress.aztec.Html;

import java.util.Map;

public class PicassoImageLoader implements Html.ImageGetter {

    private Context context;
    private Map<String, com.squareup.picasso.Target> targets;

    public PicassoImageLoader(Context context, AztecText aztec) {
        this.context = context;
        this.targets = new ArrayMap<>();

        // Picasso keeps a weak reference to targets so we need to attach them to AztecText
        aztec.setTag(targets);
    }

    @Override
    public void loadImage(final String source, final Callbacks callbacks, int maxWidth) {
        Picasso picasso = Picasso.with(context);
        picasso.setLoggingEnabled(true);

        Target target = new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                callbacks.onImageLoaded(new BitmapDrawable(context.getResources(), bitmap));
                targets.remove(source);
            }

            @Override
            public void onBitmapFailed(Drawable errorDrawable) {
                callbacks.onImageLoadingFailed();
                targets.remove(source);
            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {
            }
        };

        // add a strong reference to the target until it's called or the view gets detroyed
        targets.put(source, target);

        //noinspection SuspiciousNameCombination
        picasso.load(source).resize(maxWidth, maxWidth).centerInside().onlyScaleDown().into(target);
    }
}

File: picasso-loader/src/main/java/org/wordpress/aztec/picassoloader/PicassoImageLoader.kt
package org.wordpress.aztec.picassoloader

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.DisplayMetrics
import androidx.collection.ArrayMap

import com.squareup.picasso.Picasso
import com.squareup.picasso.Target

import org.wordpress.aztec.AztecText
import org.wordpress.aztec.Html

class PicassoImageLoader(private val context: Context, aztec: AztecText) : Html.ImageGetter {

    private val targets: MutableMap<String, com.squareup.picasso.Target>

    init {
        this.targets = ArrayMap<String, Target>()

        // Picasso keeps a weak reference to targets so we need to attach them to AztecText
        aztec.tag = targets
    }

    override fun loadImage(source: String, callbacks: Html.ImageGetter.Callbacks, maxWidth: Int) {
        loadImage(source, callbacks, maxWidth, 0)
    }

    override fun loadImage(source: String, callbacks: Html.ImageGetter.Callbacks, maxWidth: Int, minWidth: Int) {
        val picasso = Picasso.with(context)
        picasso.isLoggingEnabled = true

        val target = object : Target {
            override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                bitmap?.density = DisplayMetrics.DENSITY_DEFAULT
                callbacks.onImageLoaded(BitmapDrawable(context.resources, bitmap))
                targets.remove(source)
            }

            override fun onBitmapFailed(errorDrawable: Drawable?) {
                callbacks.onImageFailed()
                targets.remove(source)
            }

            override fun onPrepareLoad(placeHolderDrawable: Drawable?) {
                callbacks.onImageLoading(placeHolderDrawable)
            }
        }

        // add a strong reference to the target until it's called or the view gets destroyed
        targets.put(source, target)

        picasso.load(source).resize(maxWidth, maxWidth).centerInside().onlyScaleDown().into(target)
    }
}
